### ISSUE TITLE
Task subscription throws error on massive workflow instance creation

### ISSUE THREAD

**Issue by JoHeinem** [zeebe-io/zeebe#448]
* given:
  * I start the zeebe broke version 0.2.0
  * I deploy the following process: 
[extended-order-process.bpmn.zip](https://github.com/zeebe-io/zeebe/files/1298602/extended-order-process.bpmn.zip)
  * I try to start 1 million workflow instances of that process *asynchronously*
* when:
  * I open up a task subscription for "reserveOrderItems": see [here](https://pastebin.com/t0HUcaru)
* then:
  * I get either [this](https://pastebin.com/ujFXiVMh) or [that](https://pastebin.com/FgQe0nzT) error on the broker side
  * No task is completed
* expected:
  * works without any errors

**Comment by JoHeinem** [comment]
Probably an error in https://github.com/zeebe-io/zb-util/tree/master/zb-map

**Comment by Zelldon** [comment]
After running https://github.com/zeebe-io/zeebe/issues/447 I run the test with async execution.
Seems that the asynchronous method works better. The broker has written 20 GB of data until I canceled the execution.

I saw many exception for example:
```java

java.lang.RuntimeException: Could not deserialize object. Deserialization stuck at offset 4
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:46) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.LoggedEventImpl.readValue(LoggedEventImpl.java:153) ~[zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.broker.incident.processor.IncidentStreamProcessor.onWorkflowInstanceEvent(IncidentStreamProcessor.java:182) ~[classes/:?]
	at io.zeebe.broker.incident.processor.IncidentStreamProcessor.onEvent(IncidentStreamProcessor.java:141) ~[classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController$ProcessState.lambda$new$0(StreamProcessorController.java:322) ~[zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.ComposedState.doWork(ComposedState.java:63) ~[zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.StateMachineAgent.doWork(StateMachineAgent.java:53) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.doWork(StreamProcessorController.java:131) [zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:138) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: java.lang.RuntimeException: Could not read property 'payload'.
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:142) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:42) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	... 15 more
Caused by: java.lang.RuntimeException: Not binary
	at io.zeebe.msgpack.spec.MsgPackReader.readBinaryLength(MsgPackReader.java:172) ~[zb-msgpack-core-0.5.0-SNAPSHOT.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.value.BinaryValue.read(BinaryValue.java:97) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.property.BaseProperty.read(BaseProperty.java:99) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:138) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:42) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	... 15 more
11:13:46.167 [actor-runner-broker-1] ERROR io.zeebe.logstreams - The log stream processor 'workflow.default-topic.1.processor.incident' failed to process event. It stop processing further events.
java.lang.RuntimeException: Could not deserialize object. Deserialization stuck at offset 4
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:46) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.LoggedEventImpl.readValue(LoggedEventImpl.java:153) ~[zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.broker.incident.processor.IncidentStreamProcessor.onWorkflowInstanceEvent(IncidentStreamProcessor.java:182) ~[classes/:?]
	at io.zeebe.broker.incident.processor.IncidentStreamProcessor.onEvent(IncidentStreamProcessor.java:141) ~[classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController$ProcessState.lambda$new$0(StreamProcessorController.java:322) ~[zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.ComposedState.doWork(ComposedState.java:63) ~[zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.state.StateMachineAgent.doWork(StateMachineAgent.java:53) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.doWork(StreamProcessorController.java:131) [zb-logstreams-0.5.0-20171114.142606-3.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) [zb-util-0.5.0-20171113.145701-8.jar:0.5.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: java.lang.RuntimeException: Could not read property 'state'.
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:142) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:42) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	... 15 more
Caused by: java.lang.RuntimeException: Not a string
	at io.zeebe.msgpack.spec.MsgPackReader.readStringLength(MsgPackReader.java:141) ~[zb-msgpack-core-0.5.0-SNAPSHOT.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.value.StringValue.read(StringValue.java:149) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.value.EnumValue.read(EnumValue.java:80) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.property.BaseProperty.read(BaseProperty.java:99) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:138) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:42) ~[zb-msgpack-value-0.5.0-20171114.142338-4.jar:0.5.0-SNAPSHOT]
	... 15 more
```

The last and weird one was:

```java
io.zeebe.util.state.NoSuchTransitionException: Cannot take transition 3 when in state io.zeebe.logstreams.processor.StreamProcessorController$FailedState@3abdcc3
	at io.zeebe.util.state.StateMachine.take(StateMachine.java:57)
	at io.zeebe.util.state.SimpleStateMachineContext.take(SimpleStateMachineContext.java:29)
	at io.zeebe.logstreams.processor.StreamProcessorController$HandleStreamProcessorFailureState.work(StreamProcessorController.java:406)
	at io.zeebe.logstreams.processor.StreamProcessorController$HandleStreamProcessorFailureState.work(StreamProcessorController.java:382)
	at io.zeebe.util.state.TransitionState.doWork(TransitionState.java:26)
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111)
	at io.zeebe.util.state.StateMachineAgent.doWork(StateMachineAgent.java:53)
	at io.zeebe.logstreams.processor.StreamProcessorController.doWork(StreamProcessorController.java:131)
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165)
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145)
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114)
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by menski** [comment]
```
java.lang.IllegalArgumentException: No bucket in buffer 0 on offset 4
	at io.zeebe.map.BucketBufferArray.removeBucket(BucketBufferArray.java:400) ~[classes/:?]
	at io.zeebe.map.ZbMapBucketMergeHelper.tryMergeOverflowBucket(ZbMapBucketMergeHelper.java:295) ~[classes/:?]
	at io.zeebe.map.ZbMapBucketMergeHelper.tryMergingBuckets(ZbMapBucketMergeHelper.java:222) ~[classes/:?]
	at io.zeebe.map.ZbMap.remove(ZbMap.java:332) ~[classes/:?]
	at io.zeebe.map.Long2BytesZbMap.remove(Long2BytesZbMap.java:74) ~[classes/:?]
	at io.zeebe.broker.task.map.TaskInstanceMap.remove(TaskInstanceMap.java:74) ~[classes/:?]
	at io.zeebe.broker.task.processor.TaskInstanceStreamProcessor$CompleteTaskProcessor.updateState(TaskInstanceStreamProcessor.java:357) ~[classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController$ProcessState.lambda$new$3(StreamProcessorController.java:348) [classes/:?]
	at io.zeebe.util.state.ComposedState$FailSafeStep.doWork(ComposedState.java:41) ~[classes/:?]
	at io.zeebe.util.state.ComposedState.doWork(ComposedState.java:63) ~[classes/:?]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) [classes/:?]
	at io.zeebe.util.state.StateMachineAgent.doWork(StateMachineAgent.java:53) [classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController.doWork(StreamProcessorController.java:125) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) [classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```

```
java.lang.IllegalStateException: must call wrap() before
	at io.zeebe.broker.workflow.map.ActivityInstanceMap.ensureRead(ActivityInstanceMap.java:144) ~[classes/:?]
	at io.zeebe.broker.workflow.map.ActivityInstanceMap.setTaskKey(ActivityInstanceMap.java:135) ~[classes/:?]
	at io.zeebe.broker.workflow.processor.WorkflowInstanceStreamProcessor$TaskCreatedProcessor.updateState(WorkflowInstanceStreamProcessor.java:921) ~[classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController$ProcessState.lambda$new$3(StreamProcessorController.java:348) [classes/:?]
	at io.zeebe.util.state.ComposedState$FailSafeStep.doWork(ComposedState.java:41) ~[classes/:?]
	at io.zeebe.util.state.ComposedState.doWork(ComposedState.java:63) ~[classes/:?]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) [classes/:?]
	at io.zeebe.util.state.StateMachineAgent.doWork(StateMachineAgent.java:53) [classes/:?]
	at io.zeebe.logstreams.processor.StreamProcessorController.doWork(StreamProcessorController.java:125) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) [classes/:?]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) [classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```

**Comment by Zelldon** [comment]
On execution of the async workflow instance creation new excetions are seen:


```java
15:05:21.466 [actor-runner-executors-2] INFO  io.zeebe.client.subscription.task - An error ocurred when handling task 210458757840. Reporting failure to broker.
io.zeebe.client.cmd.ClientCommandRejectedException: Command for event with key 210458757840 was rejected by broker (COMPLETE_REJECTED)
	at io.zeebe.client.impl.CommandRequestHandler.getResult(CommandRequestHandler.java:171) ~[classes/:?]
	at io.zeebe.client.impl.CommandRequestHandler.getResult(CommandRequestHandler.java:41) ~[classes/:?]
	at io.zeebe.client.impl.RequestController.wrap(RequestController.java:189) ~[classes/:?]
	at io.zeebe.client.impl.RequestController$ExecuteRequestState.doWork(RequestController.java:350) ~[classes/:?]
	at io.zeebe.client.impl.RequestController$ExecuteRequestState.doWork(RequestController.java:338) ~[classes/:?]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.client.impl.RequestController.doWork(RequestController.java:162) ~[classes/:?]
	at io.zeebe.client.impl.RequestManager.doWork(RequestManager.java:82) ~[classes/:?]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
15:05:21.509 [actor-runner-executors-2] INFO  io.zeebe.client.subscription.task - Could not report failure of task 210458757840 to broker. Continuing with next task
io.zeebe.client.cmd.ClientCommandRejectedException: Command for event with key 210458757840 was rejected by broker (FAIL_REJECTED)
	at io.zeebe.client.impl.CommandRequestHandler.getResult(CommandRequestHandler.java:171) ~[classes/:?]
	at io.zeebe.client.impl.CommandRequestHandler.getResult(CommandRequestHandler.java:41) ~[classes/:?]
	at io.zeebe.client.impl.RequestController.wrap(RequestController.java:189) ~[classes/:?]
	at io.zeebe.client.impl.RequestController$ExecuteRequestState.doWork(RequestController.java:350) ~[classes/:?]
	at io.zeebe.client.impl.RequestController$ExecuteRequestState.doWork(RequestController.java:338) ~[classes/:?]
	at io.zeebe.util.state.StateMachine.doWork(StateMachine.java:111) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.client.impl.RequestController.doWork(RequestController.java:162) ~[classes/:?]
	at io.zeebe.client.impl.RequestManager.doWork(RequestManager.java:82) ~[classes/:?]
	at io.zeebe.util.actor.ActorRunner.tryRunActor(ActorRunner.java:165) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.runActor(ActorRunner.java:145) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.doWork(ActorRunner.java:114) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at io.zeebe.util.actor.ActorRunner.run(ActorRunner.java:71) ~[zb-util-0.5.0-20171120.094615-16.jar:0.5.0-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```

**Comment by Zelldon** [comment]
Fixing:

 * https://github.com/zeebe-io/zeebe/issues/550
 * https://github.com/zeebe-io/zeebe/issues/556

Seems to solve the problems. Test with 100_000 WF instances are green. @menski We should run the test with 1 million instances.

**Comment by menski** [comment]
@Zelldon the test with 1mio wf instance and 10mio tasks finished on server 3 after 1:30h and a disk usage of:

```
[camunda@cam-int-3 integration-tests]$ du -sch target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/*
3.0M    target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/gossip
37G     target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/logs
8.0K    target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/meta
4.0M    target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/metrics
1.4G    target/test-classes/io/zeebe/broker/it/workflow/persistent-broker/snapshot
39G     total
```

**Comment by Zelldon** [comment]
Very nice! On the async request's we have no problems with the broker dying?

**Comment by menski** [comment]
this should be fixed with #596
